1. Importing libraries

In [2]:
from sklearn.metrics import accuracy_score
from sklearn import tree
import pandas as pd
import numpy as np 
import scipy.io

2. Loading data using loadmat function since it's a mat file

In [3]:
train = scipy.io.loadmat(r"train.mat")
test = scipy.io.loadmat(r"test.mat")
val = scipy.io.loadmat(r"val.mat")

3. Splitting data into features and label

In [4]:
x_train = train['features']
y_train = train['label'].flatten() 

x_val = val['features']
y_val = val['label'].flatten()

x_test = test['features']
y_test = test['label'].flatten()

#print("train set",x_train.shape,y_train.shape)
#print("val set",x_val.shape,y_val.shape)
#print("test set",x_test.shape,y_test.shape)

Problem faced: unhashable type: 'slice' , as I understood, this error occurred as I wanted to split features and label using indexing/slicing/etc.. but since it's a dict, it can't be sliced. Tried to convert mat files into csv files but failed.

4. Building Machine Learning Model

In [5]:
clf = tree.DecisionTreeClassifier(max_depth=15,random_state=24)
clf = clf.fit(x_train, y_train)

Hyperparameter: 15 is chosen since I tried both higher numbers -with no increase in accuracy- and lower numebrs -with a decrease in accuracy-

5. Validation

In [6]:
y_pred = clf.predict(x_val)
print("validation accuarcy:",accuracy_score(y_val, y_pred))

validation accuarcy: 0.8055555555555556


6. Testing

In [7]:
y_pred = clf.predict(x_test)
print("testing accuarcy:",accuracy_score(y_test, y_pred))

testing accuarcy: 0.8444444444444444


Testing accuracy: 0.8444
Required accuracy: 0.95

Decision Tree won't give the required accuracy